tsl/src/remote/copy_fetcher.c

```diff
@@ -240,6 +241,66 @@ copy_data_check_header(StringInfo copy_data)
 	}
 }
 
+/*
+ * End the COPY after receiving EOF or canceling a query (e.g., due to a LIMIT
+ * being reached).
+ *
+ * This should be called after canceling a query, or, after reading all data,
+ * the file trailer, and getting an EOF return value.
+ */
+static void
+end_copy(CopyFetcher *fetcher)
+{
+	PGconn *conn = remote_connection_get_pg_conn(fetcher->state.conn);
+	PGresult *final_pgres = NULL;
+	PGresult *pgres = NULL;
+	/* Expect COMMAND OK if we received EOF. If the query was canceled (e.g.,
+	 * due to a LIMIT), expect FATAL_ERROR (query failed) */
+	ExecStatusType expected_status = fetcher->state.eof ? PGRES_COMMAND_OK : PGRES_FATAL_ERROR;
+	ExecStatusType received_status;
+
+	Assert(fetcher->state.open);
+
+	/* Read results until NULL */
+	while ((pgres = PQgetResult(conn)))
+	{
+		if (final_pgres == NULL)
+			final_pgres = pgres;
+		else
+			remote_result_close(pgres);
+	}
+
+	received_status = PQresultStatus(final_pgres);
+	remote_result_close(final_pgres);
+
+	if (received_status != expected_status)
+		remote_connection_elog(fetcher->state.conn, ERROR);
```

Comment:
This might give us messages like "Fatal error: success". Let's also add a line about what kind of status we were expecting and that it didn't match.